File: clients/file-reader/file_reader/smartedu/url_parser.py
# Mirrors chrome-extension/doc-extract/smartedu/url-parser.js — keep URL templates in sync.

"""Parse SmartEdu page URLs into detail API endpoints."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from urllib.parse import parse_qs, urlparse

_UUID = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)

DETAIL_HOST = "https://s-file-1.ykt.cbern.com.cn"


class SmartEduUrlKind(str, Enum):
    """Supported SmartEdu URL shapes."""

    CLASS_ACTIVITY = "class_activity"
    TCH_MATERIAL = "tch_material"
    PREPARE_DETAIL = "prepare_detail"
    QUALITY_COURSE = "quality_course"


@dataclass(frozen=True)
class ParsedSmartEduUrl:
    """Normalized SmartEdu URL parse result."""

    kind: SmartEduUrlKind
    resource_id: str
    detail_url: str


def _detail_path(kind: SmartEduUrlKind, resource_id: str) -> str:
    if kind == SmartEduUrlKind.CLASS_ACTIVITY:
        return f"/zxx/ndrv2/national_lesson/resources/details/{resource_id}.json"
    if kind == SmartEduUrlKind.TCH_MATERIAL:
        return f"/zxx/ndrv2/resources/tch_material/details/{resource_id}.json"
    if kind == SmartEduUrlKind.PREPARE_DETAIL:
        return f"/zxx/ndrv2/prepare_lesson/resources/details/{resource_id}.json"
    if kind == SmartEduUrlKind.QUALITY_COURSE:
        return f"/zxx/ndrv2/quality_course/resources/details/{resource_id}.json"
    raise ValueError(f"Unsupported URL kind: {kind}")


def _pick_uuid(params: dict[str, list[str]], *keys: str) -> str | None:
    for key in keys:
        values = params.get(key)
        if not values:
            continue
        candidate = values[0].strip()
        if _UUID.match(candidate):
            return candidate
    return None


def _kind_from_path(path: str) -> SmartEduUrlKind | None:
    lowered = path.lower()
    if "syncclassroom/classactivity" in lowered or "syncclassroom/class_activity" in lowered:
        return SmartEduUrlKind.CLASS_ACTIVITY
    if "tchmaterial/detail" in lowered:
        return SmartEduUrlKind.TCH_MATERIAL
    if "prepare/detail" in lowered:
        return SmartEduUrlKind.PREPARE_DETAIL
    if "qualitycourse" in lowered:
        return SmartEduUrlKind.QUALITY_COURSE
    return None


def parse_smartedu_url(raw_url: str) -> ParsedSmartEduUrl:
    """Parse a SmartEdu URL and return the detail JSON endpoint."""
    text = raw_url.strip()
    if not text:
        raise ValueError("URL is empty")

    parsed = urlparse(text)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError("URL must include scheme and host")

    kind = _kind_from_path(parsed.path)
    if kind is None:
        raise ValueError("Unsupported SmartEdu URL path")

    params = parse_qs(parsed.query, keep_blank_values=False)
    resource_keys = {
        SmartEduUrlKind.CLASS_ACTIVITY: ("activityId", "activityid"),
        SmartEduUrlKind.TCH_MATERIAL: ("contentId", "contentid", "id"),
        SmartEduUrlKind.PREPARE_DETAIL: ("lessonId", "lessonid", "contentId", "contentid"),
        SmartEduUrlKind.QUALITY_COURSE: ("courseId", "courseid", "teachingmaterialId", "contentId", "contentid"),
    }
    resource_id = _pick_uuid(params, *resource_keys[kind])
    if resource_id is None:
        raise ValueError(f"Missing resource id query parameter for {kind.value}")

    detail_path = _detail_path(kind, resource_id)
    return ParsedSmartEduUrl(
        kind=kind,
        resource_id=resource_id,
        detail_url=f"{DETAIL_HOST}{detail_path}",
    )

File: chrome-extension/doc-extract/smartedu/url-parser.js
/**
 * SmartEdu URL parser — mirrors clients/file-reader/file_reader/smartedu/url_parser.py.
 * Keep URL templates in sync with the file-reader SmartEdu tab.
 * @fileoverview Map SmartEdu page URLs to detail JSON API templates.
 */
(function (global) {
  "use strict";

  const MindGraphDocExtract = global.MindGraphDocExtract || {};

  /** @typedef {{ kind: string, activityId: string, detailUrl: string }} SmartEduParsedUrl */

  const DETAIL_TEMPLATES = {
    class_activity:
      "https://s-file-1.ykt.cbern.com.cn/zxx/ndrv2/national_lesson/resources/details/{activityId}.json",
    national_lesson:
      "https://s-file-1.ykt.cbern.com.cn/zxx/ndrv2/national_lesson/resources/details/{activityId}.json",
    tch_material:
      "https://s-file-1.ykt.cbern.com.cn/zxx/ndrs/resources/tch_material/details/{activityId}.json",
    prepare_detail:
      "https://s-file-1.ykt.cbern.com.cn/zxx/ndrs/prepare/detail/{activityId}.json",
  };

  /**
   * @param {string} pageUrl
   * @returns {SmartEduParsedUrl | null}
   */
  function parseSmartEduUrl(pageUrl) {
    if (!pageUrl || typeof pageUrl !== "string") {
      return null;
    }
    let parsed;
    try {
      parsed = new URL(pageUrl);
    } catch {
      return null;
    }
    if (!parsed.hostname.toLowerCase().includes("smartedu.cn")) {
      return null;
    }
    const activityId =
      parsed.searchParams.get("activityId") ||
      parsed.searchParams.get("contentId") ||
      parsed.searchParams.get("resourceId") ||
      "";
    if (!activityId) {
      return null;
    }
    const path = parsed.pathname.toLowerCase();
    let kind = "class_activity";
    if (path.includes("tchmaterial") || path.includes("tch_material")) {
      kind = "tch_material";
    } else if (path.includes("prepare")) {
      kind = "prepare_detail";
    } else if (path.includes("classactivity") || path.includes("syncclassroom")) {
      kind = "class_activity";
    }
    const template = DETAIL_TEMPLATES[kind] || DETAIL_TEMPLATES.class_activity;
    return {
      kind,
      activityId,
      detailUrl: template.replace("{activityId}", encodeURIComponent(activityId)),
    };
  }

  MindGraphDocExtract.parseSmartEduUrl = parseSmartEduUrl;
  MindGraphDocExtract.SMARTEDU_DETAIL_TEMPLATES = DETAIL_TEMPLATES;
  global.MindGraphDocExtract = MindGraphDocExtract;
})(typeof self !== "undefined" ? self : globalThis);
